### PR TITLE
fixes #33434  :  [Painless] Augmentation.join can't handle empty strings at the start

### LINKED ISSUE #33434: [Painless] Augmentation.join can't handle empty strings at the start
**Elasticsearch version** (`bin/elasticsearch --version`):

6.3.1

The issue is not exclusive to the mentioned version, as the code is faulty on master too.

Source links relevant:
https://github.com/elastic/elasticsearch/blob/v6.3.1/modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java#L174
https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java#L174

**Plugins installed**:

`[]` (none)

**JVM version** (`java -version`):

zulu 1.8.0_172 build of java8

**OS version** (`uname -a` if on a Unix-like system):

`3.10.0-693.el7.x86_64 #1 SMP Thu Jul 6 19:56:57 EDT 2017 x86_64 x86_64 x86_64 GNU/Linux`
But this has been tested on Debian, Ubuntu, CentOS, and other Redhat machines.

**Description of the problem including expected versus actual behavior**:

The `join` method for `Iterable`s provided by the `Augmentation` class for Painless can't deal with iterables containing zero-length members at the head of the iterable.
This script: `"script": "['', 0, '', 1, null, 3, ''].join(';')"` should output
`";0;;1;null;3;"`
But it outputs
`"0;;1;null;3;"`

The empty string on the start is appended to the StringBuilder, but the second pass will see it as `sb.length() == 0` still, and then will continue without a separator until it finds a non-zero-length member.

As counter-examples of why it should be different, the following cases output the expected one:
- In groovy: `['', 0, '', 1, null, 3, ''].join(';')`
- In java8: `String.join(";", Arrays.asList("", "0", "", "1", "null", "3", ""))`
- In java8: `Stream.of("", "0", "", "1", "null", "3", "").collect(Collectors.joining(";"))`
- In javascript: `['', 0, '', 1, null, 3, ''].join(';')`

**Steps to reproduce**:

 1. Pick literally any elasticsearch cluster with N>0 indexes/types
 2. Run the following query:
```json
GET _search
{
  "size": 1,
  "script_fields": {
    "bacon": {
      "script": "['', 0, '', 1, null, 3, ''].join(';')"
    }
  }
}

```
 3. Taste the bacon

### PR DESCRIPTION
**Pull request by nagads**

### CODE REVIEW

**Review by nik9000 — APPROVED** [review]
I think this is correct. It's been broken for so long I wonder if folks rely on it though? OTOH, we're the only ones that join like this. So I think we should merge this and backport to 7.12. @rjernst and @jdconrad, what do you think?

**Comment by nik9000** [comment]
@elasticmachine, test this please

**Comment by rjernst** [comment]
Seems fine to me, though we should fix the formatting.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java [hunk 1]
```diff
@@ -181,8 +181,13 @@ public static String namedGroup(Matcher receiver, String name) {
      */
     public static <T> String join(Iterable<T> receiver, String separator) {
         StringBuilder sb = new StringBuilder();
+        boolean firstToken = false;
```
Comment (nik9000) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:184]: @nagads, could you hit this with the formatter? It should put the `else` on the same line and remove the empty line. At some point we're going to make this a requirement for everything just so you don't have to think too much about formatting when hacking on ES. But we haven't yet.

**Comment by nik9000** [comment]
@elasticmachine, run elasticsearch-ci/1

That failure looks like a network burble.

**Comment by nik9000** [comment]
@elasticmachine, run elasticsearch-ci/1

**Comment by nik9000** [comment]
@elasticmachine, run elasticsearch-ci/bwc

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java (hunk 1 above)
Comment (nagads) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java:184]: @nik9000 thanks for kind feedback, please find the format fix in latest commit.

**Comment by nik9000** [comment]
@elasticmachine, test this please

**Comment by nik9000** [comment]
Hi @nagads ! The build is upset about the tabs and I *think* the initial value isn't quite right on the boolean. Could you run `./gradlew -pmodules/lang-painless check` to run the build? It'll spit out something about the tabs and the test.

**Comment by mark-vieira** [comment]
@elasticmachine update branch

**Comment by nagads** [comment]
> Hi @nagads ! The build is upset about the tabs and I _think_ the initial value isn't quite right on the boolean. Could you run `./gradlew -pmodules/lang-painless check` to run the build? It'll spit out something about the tabs and the test.

@nik9000 thanks for the feedback, fixed tab and boolean initialization. Ran gradlew -pmodules/lang-painless check showed build success with no errors, for the latest commit https://github.com/elastic/elasticsearch/commit/d2372e0061ad09672e2c037939673d9a0c527691

**Comment by nik9000** [comment]
Thanks @nagads! Sorry for all the iterations.

@elatsicmachine, test this please.

**Comment by nik9000** [comment]
Thanks @nagads ! I've merged to master and will backport soon!

**Comment by nik9000** [comment]
All backported!

**Comment by nagads** [comment]
@nik9000 thanks for all your help :)

**Pull request merged**